src/libstore/machines.cc

```diff
@@ -6,29 +6,24 @@
 
 namespace nix {
 
-Machine::Machine(decltype(storeUri) storeUri,
+Machine::Machine(
+    const std::string & storeUri,
     decltype(systemTypes) systemTypes,
     decltype(sshKey) sshKey,
     decltype(maxJobs) maxJobs,
     decltype(speedFactor) speedFactor,
     decltype(supportedFeatures) supportedFeatures,
     decltype(mandatoryFeatures) mandatoryFeatures,
     decltype(sshPublicHostKey) sshPublicHostKey) :
-    storeUri(
-        // Backwards compatibility: if the URI is schemeless, is not a path,
-        // and is not one of the special store connection words, prepend
-        // ssh://.
-        storeUri.find("://") != std::string::npos
-        || storeUri.find("/") != std::string::npos
-        || storeUri == "auto"
-        || storeUri == "daemon"
-        || storeUri == "local"
-        || hasPrefix(storeUri, "auto?")
-        || hasPrefix(storeUri, "daemon?")
-        || hasPrefix(storeUri, "local?")
-        || hasPrefix(storeUri, "?")
-        ? storeUri
-        : "ssh://" + storeUri),
+    storeUri([&storeUri]{
+        try {
+            return StoreReference::parse(storeUri);
+        } catch (UsageError &) {
+            // Backwards compatibility: if the URI is invalid, it might
+            // be bare host name.
+            return StoreReference::parse("ssh://" + storeUri);
```

Comment:
If this fails, we should throw the error for the unmodified string, not the new one.

Reply:
I just restored the old method of compat for now. We can decide separately whether such try-catch is worth it.